go-controller/pkg/node/node.go

```diff
@@ -342,3 +350,146 @@ func buildEndpointAddressMap(epSubsets []kapi.EndpointSubset) map[epAddressItem]
+		AddFunc: func(obj interface{}) {
+			klog.Infof("AddFunc:")
+			pod := obj.(*kapi.Pod)
+			if !ovn.PodWantsNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
+				return
+			}
+			if ovn.PodScheduled(pod) {
+				// Is this pod created on same node where the smart NIC
+				if n.name != pod.Spec.NodeName {
+					return
+				}
+
+				pfindex, vfindex, ok := n.getSmartNicAnnotations(pod)
+				if !ok {
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				// get POD annotation MAC/IP/GW/
+				podInfo, err := util.UnmarshalPodAnnotation(pod.Annotations)
+				if err != nil {
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				err = n.addRepPort(pod, pfindex, vfindex, podInfo)
+				if err != nil {
+					retryPods.Store(pod.UID, true)
+				} else {
+					servedPods.Store(pod.UID, true)
+				}
+			} else {
+				// Handle unscheduled pods later in UpdateFunc
+				retryPods.Store(pod.UID, true)
+				return
+			}
+		},
+		UpdateFunc: func(old, newer interface{}) {
+			klog.Infof("UpdateFunc:")
+			pod := newer.(*kapi.Pod)
+			if !ovn.PodWantsNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
+				retryPods.Delete(pod.UID)
+				return
+			}
+			_, retry := retryPods.Load(pod.UID)
+			if ovn.PodScheduled(pod) && retry {
+				if n.name != pod.Spec.NodeName {
+					retryPods.Delete(pod.UID)
+					return
+				}
+				pfindex, vfindex, ok := n.getSmartNicAnnotations(pod)
+				if !ok {
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				// get POD annotation MAC/IP/GW/
+				podInfo, err := util.UnmarshalPodAnnotation(pod.Annotations)
+				if err != nil {
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				err = n.addRepPort(pod, pfindex, vfindex, podInfo)
+				if err != nil {
+					retryPods.Store(pod.UID, true)
+				} else {
+					servedPods.Store(pod.UID, true)
+					retryPods.Delete(pod.UID)
+				}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			klog.Infof("DeleteFunc:")
+			pod := obj.(*kapi.Pod)
+			if _, ok := servedPods.Load(pod.UID); !ok {
+				return
+			}
+			servedPods.Delete(pod.UID)
+			pfIndex, vfIndex, ok := n.getSmartNicAnnotations(pod)
+			if !ok {
+				return
+			}
+			_ = n.delRepPort(pod, pfIndex, vfIndex)
+		},
+	}, nil)
+	return err
+}
+
+// getSmartNicAnnotations return the smart nic annotation which inclue the PF and VF indexes
+func (n *OvnNode) getSmartNicAnnotations(pod *kapi.Pod) (string, string, bool) {
+	pfIndex, ok1 := pod.Annotations["ovn.smartnic.pf"]
+	vfIndex, ok2 := pod.Annotations["ovn.smartnic.vf"]
+	isFound := ok1 && ok2
+	return pfIndex, vfIndex, isFound
+}
+
+// addRepPort adds the representor of the VF to the ovs bridge
+func (n *OvnNode) addRepPort(pod *kapi.Pod, pfindex, vfindex string, ifInfo *util.PodAnnotation) error {
+	return wait.PollImmediate(500*time.Millisecond, 60*time.Second, func() (bool, error) {
+		portName := fmt.Sprintf("pf%svf%s", pfindex, vfindex)
+		ifaceID := fmt.Sprintf("%s_%s", pod.Namespace, pod.Name)
```

Comment:
This will not work with the newly proposed scheme for VF representors in SmartNICs

```
Details:
-----------
Current state:
-------------------
1.	What we have today in BF(smartnic)?
a.	PF representor phys_port_name=pf0
b.	VF representor phys_port_name=pf0vf0

What is missing?
1.	This naming scheme is unable to define multi-host.
A given representor is for which external host controller is missing.
Hence, we need non ambiguous definition of which host controller this belongs to.
Though switch_id may be good enough to match, but switch_id is an not a right field to define external host controller(s).

2.	When a user wants to create an SF, user can create a SF either on the BF ARM PF0 or x86 PF0.
So user telling below command doesn’t tell whether to create an SF on BF PF0 or external host controller PF 0.
Both the host controllers have same switch id.
$ devlink port add pci/0000:03:00.0 flavour pcisf pfnum 0 sfnum 10
Hence, we need a non-ambiguous definition of host controller number to create.

To address above two requirements, phys_port_name is changing hopefully for last time.
So what is changing:
1.	A representor phys_port_name will have additional “cN” to identify external host controller.
“pf0” -> “c0pf0”
“pf0vf0” -> “c0pf0vf0”
2.	Devlink example
$ devlink port show pci/0000:00:08.0/2
pci/0000:00:08.0/2: type eth netdev eth7 controller 0 flavour pcivf pfnum 0 vfnum 1 splittable false
  function:
    hw_addr 00:00:00:00:00:00

$ devlink port show -jp pci/0000:00:08.0/2 {
    "port": {
        "pci/0000:00:08.0/1": {
            "type": "eth",
            "netdev": "eth7",
            "controller": 0,    /* new addition */
            "flavour": "pcivf",
            "pfnum": 0,
           "vfnum": 1,
            "splittable": false,
            "function": {
                "hw_addr": "00:00:00:00:00:00"
            }
        }
    }
}

3.	systemd/udev example:
An example representor netdev udev name consist of controller annotation for external controller with controller number = 0, for PF 0 and VF 1:
$ udevadm test-builtin net_id /sys/class/net/eth7 Using default interface naming scheme 'v245'.
ID_NET_NAMING_SCHEME=v245
ID_NET_NAME_PATH=enp0s8f0nc0pf0vf1
Unload module index
Unloaded link configuration context.
```